src/main/java/com/github/webicitybrowser/webicity/renderer/backend/html/tags/TagAction.java

```diff
@@ -0,0 +1,36 @@
+package com.github.webicitybrowser.webicity.renderer.backend.html.tags;
+
+import com.github.webicitybrowser.spec.css.parser.CSSParser;
+import com.github.webicitybrowser.spec.css.parser.tokenizer.CSSTokenizer;
+import com.github.webicitybrowser.spec.css.parser.tokens.Token;
+import com.github.webicitybrowser.spec.css.rule.CSSRule;
+import com.github.webicitybrowser.spec.css.rule.CSSRuleList;
+import com.github.webicitybrowser.spec.dom.node.Element;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.util.HashMap;
+
+public interface TagAction {
+
+	CSSRuleList getCSSRuleList(Element element);
+
+	default CSSRuleList createRuleList(Reader childContextReader) {
+		try {
+			Token[] tokens = CSSTokenizer.create().tokenize(childContextReader);
+			CSSRule[] rules = CSSParser.create().parseAListOfRules(tokens);
+			return CSSRuleList.create(rules);
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	static TagAction getAction(String elementName) {
```

Comment:
Since you are populating the HashMap for every single element popped off the HTML stack, this is probably not great for performance

Instead of this method, I would make an interface called TagActions, whose implementation is HashMap-backed, and that has various methods like `registerTagAction`. The implementation should just be instantiated before the `HTMLRendererBackendParserSettings` is created, and passed to that.

I would use TagActions for other things in the future, so we would have multiple instances of it